cluster-autoscaler/proposals/granular-resource-limits.md

```diff
@@ -0,0 +1,328 @@
+the namespace in which they are defined, while the nodes are global. We can’t
+use namespaced resources to limit the creation and deletion of global resources.
+
+## User Stories
+
+### Story 1
+
+As a cluster administrator, I want to configure cluster-wide resource limits to
+avoid excessive cloud provider costs.
+
+**Note:** This is already supported in CAS, but not in Karpenter.
+
+Example AutoscalingResourceQuota:
+
+```yaml
+apiVersion: autoscaling.x-k8s.io/v1beta1
+kind: AutoscalingResourceQuota
+metadata:
+  name: cluster-wide-limits
+spec:
+  limits:
+    resources:
+      cpu: 128
+      memory: 256Gi
+```
+
+### Story 2
+
+As a cluster administrator, I want to configure separate resource limits for
+specific groups of nodes on top of cluster-wide limits, to avoid a situation
+where one group of nodes starves others of resources.
+
+**Note:** A specific group of nodes can be either a NodePool in Karpenter, a
+ComputeClass in GKE, or simply a set of nodes grouped by a user-defined label.
+This can be useful e.g. for organizations where multiple teams are running
+workloads in a shared cluster, and these teams have separate sets of nodes. This
+way, a cluster administrator can ensure that each team has a proper limit for
+their resources and it doesn’t starve other teams. This story is partly
+supported by Karpenter’s NodePool limits.
+
+Example AutoscalingResourceQuota:
+
+```yaml
+apiVersion: autoscaling.x-k8s.io/v1beta1
+kind: AutoscalingResourceQuota
+metadata:
+  name: team-a-limits
+spec:
+  selector:
+    matchLabels:
+      team: a
+  limits:
+    resources:
+      cpu: 32
+```
+
+### Story 3
+
+As a cluster administrator, I want to allow scaling up machines that are more
+expensive or less suitable for my workloads when better machines are
+unavailable, but I want to limit how many of them can be created, so that I can
+control extra cloud provider costs, or limit the impact of using non-optimal
+machine for my workloads.
+
+Example AutoscalingResourceQuota:
+
+```yaml
+apiVersion: autoscaling.x-k8s.io/v1beta1
+kind: AutoscalingResourceQuota
+metadata:
+  name: max-e2-resources
+spec:
+  selector:
+    matchLabels:
+      example.cloud.com/machine-family: e2
+  limits:
+    resources:
+      cpu: 32
+      memory: 64Gi
+```
+
+### Story 4
+
+As a cluster administrator, I want to limit the number of nodes in a specific
+zone if my cluster is unbalanced for any reason, so that I can avoid exhausting
+IP space in that zone, or enforce better balancing across zones.
+
+**Note:** Originally requested
+in [https://github.com/kubernetes/autoscaler/issues/6940](https://github.com/kubernetes/autoscaler/issues/6940).
+
+Example AutoscalingResourceQuota:
+
+```yaml
+apiVersion: autoscaling.x-k8s.io/v1beta1
+kind: AutoscalingResourceQuota
+metadata:
+  name: max-nodes-us-central1-b
+spec:
+  selector:
+    matchLabels:
+      topology.kubernetes.io/zone: us-central1-b
+  limits:
+    resources:
+      nodes: 64
```

Comment:
Did you look at how ResourceQuotas do this? https://kubernetes.io/docs/concepts/policy/resource-quotas/#quota-on-object-count

```
spec:
  hard:
    cpu: 100
    count/nodes: 3
```

Reply:
Naming it just `count` might be not descriptive enough, as there are limits on other resources too. However, `capacityresourcequota.spec.hard.nodes` makes sense to me. I guess in such case it would look like this:
```yaml
spec:
  limits:
    nodes: 3
    resources:
      cpu: 16
```

The question is whether moving `nodes` outside of `resources` improves this API in any way. I'd probably stick with keeping it in `resources` to keep it simple, but I won't insist on that if you have a strong preference